document/FileFormat.mdk

```diff
@@ -2696,31 +2703,318 @@ The tool might provide a "related location" (&sect;[#result-relatedLocations]) w
+|                  | name of the variable        | declaration has an          | declaration has an          |
+|                  | being declared. If          | initializer, in which       | initializer, in which case, |
+|                  | multiple variables are      | case, the value of          | the names and values of     |
+|                  | declared in the same        | the initializer             | any variables participating |
+|                  | statement, a separate       | expression.                 | in the initializer          |
+|                  | `annotatedCodeLocation`     | Absent: If the              | expression.                 |
+|                  | object shall be created     | declaration has no          | Absent: If the              |
+|                  | for each declaration that   | initializer.                | declaration has no          |
+|                  | the tool creating the       |                             | initializer.                |
+|                  | log wishes to represent.    |                             |                             |
+|------------------|-----------------------------|-----------------------------|-----------------------------|
+| `functionEnter`  | Recommended: The            | Absent                      | Absent                      |
+|                  | fully qualified name of     |                             |                             |
+|                  | the function being          |                             |                             |
+|                  | entered.                    |                             |                             |
+|------------------|-----------------------------|-----------------------------|-----------------------------|
+| `functionExit`   | Recommended: The            | Optional: If the function   | Absent                      |
+|                  | fully qualified name of     | returns a value or          |                             |
+|                  | the function being          | values, in which case,      |                             |
+|                  | returned from.              | string representations      |                             |
+|                  |                             | of the value or values      |                             |
+|                  |                             | returned.                   |                             |
+|                  |                             | Absent: If the function     |                             |
+|                  |                             | does not return a value.    |                             |
+|------------------|-----------------------------|-----------------------------|-----------------------------|
+| `functionReturn` | Recommended: The            | Optional: If the function   | Optional: If the function   |
```

Comment:
Per our discussion: Would you prefer this to be "optional" if the `functionReturn` _is_ nested within a `functionEnter/functionExit` pair, or do you want to prohibit it ("absent")? I can see it both ways:

* In favor of "optional":
    1. What's the harm of having it?
    2. We wouldn't have to add a smart validator to ensure that it's _not_ there if the `aCL` _is_ within a `functionEnter/functionExit` pair.

* In favor of "absent":
    1. Slightly smaller file size (IMO not significant)
    2. Avoids possible inconsistency with the name as specified in the `functionReturn` and the name as specified in the `functionExit`.

I'd make it optional.

---
In reply to: [77425829](https://github.com/sarif-standard/sarif-spec/pull/264#discussion_r77425829) [](ancestors = 77425829)